src/storage/transaction/TransactionManager.cpp

```diff
@@ -0,0 +1,567 @@
+    VertexIDSlice dst = NebulaKeyUtils::getDstId(vIdLen, rawKey);
+    EdgeType type = NebulaKeyUtils::getEdgeType(vIdLen, rawKey);
+    EdgeRanking rank = NebulaKeyUtils::getRank(vIdLen, rawKey);
+
+    auto inPartIdStatus = env_->metaClient_->partId(spaceId, src.str());
+    CHECK(inPartIdStatus.ok());
+    auto inPartId = inPartIdStatus.value();
+
+    return NebulaKeyUtils::edgeKey(vIdLen, inPartId, dst.str(), 0 - type, rank, src.str(), ver);
+}
+
+folly::Future<cpp2::ErrorCode> TransactionManager::resumeTransaction(size_t vIdLen,
+                                                                     GraphSpaceID spaceId,
+                                                                     std::string lockKey,
+                                                                     std::string lockVal,
+                                                                     ResumedResult result) {
+    std::string rawKey = NebulaKeyUtils::edgeKey(lockKey);
+    CHECK(NebulaKeyUtils::isEdge(vIdLen, rawKey));
+    auto spCode = std::make_shared<cpp2::ErrorCode>(cpp2::ErrorCode::SUCCEEDED);
+
+    int64_t ver = NebulaKeyUtils::getVersion(vIdLen, rawKey);
+    auto keyWoVer = NebulaKeyUtils::keyWithNoVersion(rawKey).str();
+
+    auto lockResult = memLock_.insert(std::make_pair(keyWoVer, ver));
+    if (!lockResult.second) {
+        return folly::makeFuture(cpp2::ErrorCode::E_ADD_EDGE_CONFLICT);
+    }
+
+    auto c = folly::makePromiseContract<cpp2::ErrorCode>();
+    auto cbGetValue =
+        [=, key = rawKey](ErrOrVal&& errOrStr) mutable -> folly::SemiFuture<cpp2::ErrorCode> {
+        if (!nebula::ok(errOrStr)) {
+            if (nebula::error(errOrStr) == cpp2::ErrorCode::E_KEY_NOT_FOUND) {
+                return cpp2::ErrorCode::SUCCEEDED;
+            }
+            return nebula::error(errOrStr);
+        }
+        auto lockedPtr = result->wlock();
+        if (!lockedPtr) {
+            return cpp2::ErrorCode::E_RESUME_EDGE_CONFLICT;
+        }
+
+        auto& kv = *lockedPtr;
+        if (FLAGS_enable_multi_versions) {
+            auto verPrev = NebulaKeyUtils::isEdge(vIdLen, kv.first)
+                               ? folly::Endian::big(NebulaKeyUtils::getVersion(vIdLen, kv.first))
+                               : std::numeric_limits<int64_t>::max();
+            auto verCurr = folly::Endian::big(ver);
+            LOG_IF(INFO, FLAGS_trace_toss)
+                << "ver = " << ver << ", verCurr = " << verCurr << ", verPrev = " << verPrev;
+
+            if (verCurr > verPrev) {
+                LOG_IF(INFO, FLAGS_trace_toss) << "failed to update edge from verPrev = " << verPrev
+                                               << ", to verCurr = " << verCurr;
+                return cpp2::ErrorCode::E_RESUME_EDGE_CONFLICT;
+            }
+        }
+
+        kv.first = rawKey;
+        kv.second = nebula::value(errOrStr);
+
+        auto partId = NebulaKeyUtils::getPart(rawKey);
+        if (lockVal.empty()) {
```

Comment:
Better to check `lockVal`, I have not found a way that it is empty during review.

Reply:
good catch & fixed